src/umpire/resource/DefaultMemoryResource.inl

```diff
@@ -55,8 +56,15 @@ void DefaultMemoryResource<_allocator>::deallocate(void* ptr)
 
   UMPIRE_RECORD_STATISTIC(getName(), "ptr", reinterpret_cast<uintptr_t>(ptr), "size", 0x0, "event", "deallocate");
 
-  deregisterAllocation(ptr);
+  auto record = deregisterAllocation(ptr);
+
+  if (record->m_strategy != this) {
+    UMPIRE_ERROR(ptr << " was not allocated by " << getName());
+  }
+
```

Comment:
I really like this!  Since this pattern repeats, I wonder if it would be better in a single helper function?

Reply:
Almost - the difference is due to the error messages. In the AllocationTracker case, since that's not really exposed to the user I pull the underlying allocator name to generate a better error message.

Reply:
Then may I suggest that you pass in the string of the error message as an argument or throw an exception that the callers may catch?

Reply:
😄